src/R/Core/Impl/DataTips/RDataTip.cs

```diff
@@ -0,0 +1,114 @@
+                            return null;
+                    }
+                } else {
+                    var tokenNode = node as TokenNode;
+                    if (tokenNode == null) {
+                        return null;
+                    }
+
+                    switch (tokenNode.Token.TokenType) {
+                        case RTokenType.OpenSquareBracket:
+                        case RTokenType.CloseSquareBracket:
+                        case RTokenType.OpenDoubleSquareBracket:
+                        case RTokenType.CloseDoubleSquareBracket:
+                            break;
+
+                        case RTokenType.Operator: {
+                                string op = ast.Root.TextProvider.GetText(tokenNode.Token);
+                                switch (op) {
+                                    case "$":
+                                    case "@":
+                                    case "::":
+                                    case ":::":
+                                        break;
+                                    default:
+                                        return null;
+                                }
+                                break;
+                            }
+
+                        default:
+                            return null;
+                    }
+                }
+            }
+
+            // When the lookup starts at [ or [[, the immediate node is the token, but its parent is an Indexer node,
+            // and the parent of the indexer is the Operator. The loop below assumes that Operator is the immediate
+            // parent, so walk one level up before entering it in this case.
+            var indexer = node.Parent as Indexer;
+            if (indexer != null) {
+                node = indexer;
+            }
+
+            // Walk the AST tree up to determine the expression that should be evaluated, according to the following rules:
+            // - if current node is a child of $, @, :: or ::: (on either side of the operator), keep walking;
+            // - if current node is a left child of [ or [[, keep walking;
+            // - otherwise, stop.
+            // Thus, for an expression like a::b$c@d, the entire expression will be considered when hovering over a, b, c or d.
+            // But for a[b$c][[d]], hovering over a will consider a[b$c][d], but hovering over b or c will only consider b$c.
+            while (true) {
+                var parent = node.Parent as Operator;
+                if (parent == null) {
+                    break;
+                }
+
+                var op = parent.OperatorType;
+                if (op == OperatorType.Index) {
+                    if (parent.LeftOperand != node) {
+                        break;
+                    }
+                } else if (op != OperatorType.ListIndex && op != OperatorType.Namespace) {
+                    break;
+                }
+
+                node = parent;
+            }
```

Comment:
Perhaps it's better to refactor the larger chunk that validates the initial node?